lib/swig/base_model/model.i

```diff
@@ -7,18 +7,50 @@
 
 // We need to expose this empty class to Python, so that we can pass
 // shared pointers of it to models
-class Model {
 
+template <class T>
+class TModel{
  public:
+  TModel(){}
+  virtual void grad(const Array<T>& coeffs, Array<T>& out);
+  virtual T loss(const Array<T>& coeffs);
+  virtual unsigned long get_epoch_size() const;
+  virtual bool is_sparse() const;  
+};
 
-  Model() { }
-
+%rename(Model) TModel<double>;
+class TModel<double>{
+ public:
+  Model(){}
   virtual void grad(const ArrayDouble& coeffs, ArrayDouble& out);
   virtual double loss(const ArrayDouble& coeffs);
-
   virtual unsigned long get_epoch_size() const;
+  virtual bool is_sparse() const;  
+};
+typedef TModel<double> Model;
+typedef std::shared_ptr<Model> ModelPtr;
 
-  virtual bool is_sparse() const;
+%rename(ModelDouble) TModel<double>;
```

Comment:
Will we need to duplicate swig code for all model classes?

Reply:
Or can we do something like
```
%template(ModelLabelsFeatures) TModelLabelsFeatures<double>;
typedef TModelLabelsFeatures<double> ModelLabelsFeatures;
```

?

Reply:
This is tricky - I have found that %tempate only works if you have no signatures with templates

i.e. most constructors have references to templated array types, we get type errors unless we specify the whole class using the typedefs "ArrayDouble" instead of "Array<double>"

Reply:
This is really not a good news, I'm afraid that we will end up with methods only added to double... 
Do you think there is a way to do this with a macro?

Reply:
with proper unit tests for all types I don't see the problem